### ISSUE TITLE
Receiving messages only when subscribing with wildcard

### ISSUE THREAD

**Issue by elisiariocouto** [erlio/vernemq#94]
Hello all.
I have a three-node VerneMQ cluster (0.12.4) running for a few weeks without any problem.
All of a sudden, I started to receiving messages only when I subscribe with a wildcard in the topic.
My topic tree is like this:
`iot/<type_of_sensor>/<geo_id>/<sensor_id>`

If I subscribe to `iot/temp/1000/+` I get nothing.
If I subscribe to `iot/+/1000/+` I have data. And yes, the first topic should return exactly the same thing as the second one, on my case.

The strangest thing is that this error only happened on a specific `geo_id`. The other `geo_ids` I have, works flawlessly.

Here are some crash reports. The log files are full of this.
```
2016-03-03 17:08:35 =CRASH REPORT====
  crasher:
    initial call: vmq_ranch:init/4
    pid: <0.19183.0>
    registered_name: []
    exception error: {{badmatch,{error,enotconn}},[{vmq_ranch,init,4,[{file,"/vernemq/distdir/0.12.4/_build/default/lib/vmq_server/src/vmq_ranch.erl"},{line,57}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,237}]}]}
    ancestors: [<0.1683.0>,<0.1682.0>,ranch_sup,<0.58.0>]
    messages: []
    links: [<0.1683.0>,#Port<0.13329>]
    dictionary: []
    trap_exit: false
    status: running
    heap_size: 610
    stack_size: 27
    reductions: 142
  neighbours:
2016-03-03 17:08:35 =ERROR REPORT====
Ranch listener {{0,0,0,0},1883} had connection process started with vmq_ranch:start_link/4 at <0.19183.0> exit with reason: {badmatch,{error,enotconn}}
```
and this one
```
2016-03-03 15:55:27 =CRASH REPORT====
  crasher:
    initial call: vmq_ranch:init/4
    pid: <0.31624.259>
    registered_name: []
    exception error: bad argument: [{ets,lookup,[vmq_status,ready],[]},{vmq_cluster,is_ready,0,[{file,"/vernemq/distdir/0.12.4/_build/default/lib/vmq_server/src/vmq_cluster.erl"},{line,66}]},{vmq_reg,publish,1,[{file,"/vernemq/distdir/0.12.4/_build/default/lib/vmq_server/src/vmq_reg.erl"},{line,271}]},{vmq_mqtt_fsm,maybe_publish_last_will,1,[{file,"/vernemq/distdir/0.12.4/_build/default/lib/vmq_server/src/vmq_mqtt_fsm.erl"},{line,798}]},{vmq_mqtt_fsm,terminate,2,[{file,"/vernemq/distdir/0.12.4/_build/default/lib/vmq_server/src/vmq_mqtt_fsm.erl"},{line,418}]},{vmq_mqtt_fsm,in,3,[{file,"/vernemq/distdir/0.12.4/_build/default/lib/vmq_server/src/vmq_mqtt_fsm.erl"},{line,154}]},{vmq_mqtt_fsm,msg_in,2,[{file,"/vernemq/distdir/0.12.4/_build/default/lib/vmq_server/src/vmq_mqtt_fsm.erl"},{line,142}]},{vmq_ranch,handle_message,2,[{file,"/vernemq/distdir/0.12.4/_build/default/lib/vmq_server/src/vmq_ranch.erl"},{line,200}]}]
    ancestors: [<0.276.0>,<0.275.0>,ranch_sup,<0.58.0>]
    messages: []
    links: [<0.276.0>,#Port<0.1976473>]
    dictionary: [{random_seed,{1458,14857,27844}},{guid,{{'verne03@<IP REDACTED>',{1457,14856,797210}},5679}}]
    trap_exit: true
    status: running
    heap_size: 2586
    stack_size: 27
    reductions: 5005509
  neighbours:
```

Thanks for the help

**Comment by dergraf** [comment]
Thanks for reporting. Can you send us the complete logs over mail, please?

**Comment by ioolkos** [comment]
Is there something special only for that subscriber? a mobile phone subscribing?

(maybe related: https://github.com/erlio/vernemq/issues/41)

**Comment by elisiariocouto** [comment]
Nothing special on the subscriber nor the publisher. C's `libmosquitto` on the publisher and mosquitto_sub (1.4.7) for the subscription. I tested subscribing on multiple machines with `mosquitto_sub` - no go. I even tested subscribing from a droplet using Go's Paho library, same result.

@dergraf: Unfortunately, I don't have the logs anymore for the day I encountered this behaviour. 
I solved the problem restarting the third node. I'm still having the `{{badmatch,{error,enotconn}}` errors but the cluster seems stable. I'll send the most recent logs by e-mail.

**Comment by ioolkos** [comment]
Hm, thanks, interesting.

Maybe @ralight has seen something like that, an ENOTCONN error on only a single subscriber/topic with mosquitto_sub? We'd be very thankful if you could kindly give your opinion.

Although from @elisiariocouto's report, it seems this is entirely server based?

**Comment by elisiariocouto** [comment]
I think this is server based. It's important to notice that I have HAProxy load-balancing the connections for the three VerneMQ nodes. Maybe these ENOTCONN errors are related to this?
@dergraf, I can't find any e-mail address in your profile to send the logs

**Comment by elisiariocouto** [comment]
Another strange behaviour: If I run `vmq-admin session list` with the default options, I get this error:
```
root@verne01:~# vmq-admin session list
RPC to 'verne01@<IP_REDACTED>' failed: {'EXIT',
                                        {function_clause,
                                         [{lists,seq,
                                           [1,-1],
                                           [{file,"lists.erl"},{line,241}]},
                                          {clique_table,char_seq,2,
                                           [{file,
                                             "/vernemq/distdir/0.12.4/_build/default/lib/clique/src/clique_table.erl"},
                                            {line,196}]},
                                          {clique_table,
                                           '-vertical_border/1-lc$^0/1-0-',1,
                                           [{file,
                                             "/vernemq/distdir/0.12.4/_build/default/lib/clique/src/clique_table.erl"},
                                            {line,191}]},
                                          {clique_table,
                                           '-vertical_border/1-lc$^0/1-0-',1,
                                           [{file,
                                             "/vernemq/distdir/0.12.4/_build/default/lib/clique/src/clique_table.erl"},
                                            {line,192}]},
                                          {clique_table,vertical_border,1,
                                           [{file,
                                             "/vernemq/distdir/0.12.4/_build/default/lib/clique/src/clique_table.erl"},
                                            {line,191}]},
                                          {clique_table,create_table,4,
                                           [{file,
                                             "/vernemq/distdir/0.12.4/_build/default/lib/clique/src/clique_table.erl"},
                                            {line,71}]},
                                          {clique_human_writer,write_table,1,
                                           [{file,
                                             "/vernemq/distdir/0.12.4/_build/default/lib/clique/src/clique_human_writer.erl"},
                                            {line,67}]},
                                          {clique_human_writer,write_status,
                                           2,
                                           [{file,
                                             "/vernemq/distdir/0.12.4/_build/default/lib/clique/src/clique_human_writer.erl"},
                                            {line,57}]}]}}
```
Also, this appears to be related to `--peer_host` and `--peer_port` flags being used at the same time.
If I run `vmq-admin session list --client_id --recv_cnt --peer_host` or `vmq-admin session list --client_id --recv_cnt --peer_port`, I get the expected result. If I run `vmq-admin session list --client_id --recv_cnt --peer_host --peer_port`, I got the same error.

**Comment by ioolkos** [comment]
Thanks for reporting the vmq-admin misbehaviour.

You can mail the logs to the `info` user for the verne.mq domain.

**Comment by ralight** [comment]
There is no reason why mosquitto_sub would behave differently with a + or not, it doesn't really look at the subscription topics.

**Comment by ioolkos** [comment]
@ralight thanks a lot for this confirmation.

@elisiariocouto Yes, could be HAProxy related. It seems that a couple of mature Erlang projects have had their share of problems with enotconn and HAProxy, RabbitMQ and Riak to name a few.

But I can't quite figure out why it should be related to a single subscription/topic.

**Comment by ioolkos** [comment]
We currently think there are 3 things at work here.

1. enotconn error caused by health check of HAProxy. This is harmless, but we will fix it.

2. Errors due to constant queue migration. This might of course be made worse by a bug/bugs. As we have re-implemented queue migration, we hope this will be fixed in 0.12.5

3. The subscription error which is not understood right now. Do you use clean-session true or false in session setup?

It seems you are re-connecting your clients superfast in a round-robin manner. 
This causes the queues to migrate from cluster node to cluster node multiple times within a second.
One thing to try here is to configure HAProxy to use some sort of sticky sessions.

**Comment by elisiariocouto** [comment]
We'll have a look on HAProxy config. About the subscription error, we are using clean-session true on all our publishers and subscribers.

"It seems you are re-connecting your clients superfast in a round-robin manner."
Are you talking about the publisher side or is this HAProxy-related?
Also, we have an unreachable sensor that is trying to publish with wrong credentials almost every second. Could this be related?

**Comment by dergraf** [comment]
Thanks for your reply.
Is the 'unreachable' sensor using the client-id of another 'reachable' sensor?

**Comment by elisiariocouto** [comment]
No, it is using a random ID from `libmosquitto`. Actually, today our cluster became completely unstable. My guess is that a netsplit happened. Even after I told the "healthy" nodes that the third node was leaving, they were still trying to send data to it.

**Comment by ioolkos** [comment]
@elisiariocouto sorry to hear that. Have you seen a specific error?
I think, in general, we have to address and fix cluster leave problems as soon as possible.

**Comment by elisiariocouto** [comment]
@ioolkos, I was only seeing the same errors as before. We think that these errors we're caused by HAProxy's load-balancing feature. We're doing some more tests.

**Comment by larshesel** [comment]
Hi, since last we've published VerneMQ 0.12.5p4 and recommend upgrading. It fixes (among other things)  some cluster leave bugs.

Closing this for now, but feel free to reopen or create a new issue.
